### PR TITLE
Admin design

### PR DESCRIPTION
**Pull request by vikaluk**
**STATUS**
READY/IN DEVELOPMENT

**DESCRIPTION**

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [x] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs

### CODE REVIEW

File: app/assets/javascripts/users.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (igorkasyanchuk) [inline app/assets/javascripts/users.coffee]: remove this empty file

and other if they are not needed

File: app/controllers/account/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    @user = resource
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    binding.pry
```
Comment (igorkasyanchuk) [inline app/controllers/account/users_controller.rb]: remove binding.pry

File: app/models/device.rb [hunk 3]
```diff
@@ -1,5 +1,6 @@
 class Device < ApplicationRecord
+	belongs_to :device_group
```
Comment (igorkasyanchuk) [inline app/models/device.rb]: formatting 2 spaces (and tell it to all team members)

File: app/models/organization.rb [hunk 4]
```diff
@@ -1,4 +1,5 @@
 class Organization < ApplicationRecord
-  validates :name, presence: true, length: { maximum: 40 }
   has_many :users, dependent: :destroy
+  has_many :device_groups
```
Comment (igorkasyanchuk) [inline app/models/organization.rb]: specify dependent: :destroy

File: db/migrate/20190422154904_create_users.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class CreateUsers < ActiveRecord::Migration[5.2]
+  def change
+    create_table :users do |t|
+      t.string   :first_name
+      t.string   :last_name
+      t.string   :email
+      t.string   :password
+      t.string   :password_confirmation
+      t.integer  :role
+      t.integer  :organization_id
```
Comment (igorkasyanchuk) [inline db/migrate/20190422154904_create_users.rb]: add index by organization_id in ALL models where you have it
think about indexes

File: spec/helpers/users_helper_spec.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+#
+# describe UsersHelper do
+#   describe "string concat" do
+#     it "concats two strings with spaces" do
+#       expect(helper.concat_strings("this","that")).to eq("this that")
+#     end
+#   end
+# end
+RSpec.describe UsersHelper, type: :helper do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (igorkasyanchuk) [inline spec/helpers/users_helper_spec.rb]: remove this empty spec

File: spec/models/account/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+require 'rails_helper'
+
+RSpec.describe Account::User, type: :model do
+  context 'validation tests' do
+    let(:user) { build(:user) }
```
Comment (igorkasyanchuk) [inline spec/models/account/user_spec.rb]: always use "let!" instead of "let"

File: spec/models/device_group_spec.rb [hunk 8]
```diff
@@ -1,26 +1,25 @@
 require 'rails_helper'
 
 RSpec.describe DeviceGroup, type: :model do
-  subject {
-    described_class.new(name: "Anything", organization_id: "1")
-  }
+  context 'validation tests' do
+  let(:device_group){create (:device_group)}
```
Comment (igorkasyanchuk) [inline spec/models/device_group_spec.rb]: formatting and use "let!"

File: spec/models/device_group_spec.rb [hunk 9]
```diff
@@ -1,26 +1,25 @@
...
+  
+    it "is valid with valid params" do
+    expect(device_group.save).to eq(true)
```
Comment (igorkasyanchuk) [inline spec/models/device_group_spec.rb]: formatting
2 spaces

File: spec/models/device_spec.rb [hunk 10]
```diff
@@ -1,28 +1,25 @@
 require 'rails_helper'
 
 RSpec.describe Device, type: :model do
-  subject {
-    described_class.new(name: "Anything", device_type: "Lorem ipsum", device_group_id: "1")
-  }
+  let(:device){create (:device)}
```
Comment (igorkasyanchuk) [inline spec/models/device_spec.rb]: "let!"

File: app/assets/javascripts/spacex/main.js [hunk 11]
```diff
@@ -0,0 +1,693 @@
+(function($) {
+	"use strict";
```
Comment (romaslmd) [inline app/assets/javascripts/spacex/main.js]: should be first line in the file

File: app/assets/stylesheets/spacex/style.css [hunk 12]
```diff
@@ -0,0 +1,4779 @@
...
+51. Todo Widget
+52. X-Editable
+53. Sparkline Chart
+54. SweetAlert
+-------------------------------------------------------------------------*/
+/*-----------------------------------------------------------------------
+0. Import file styles
+*/
+@import 'https://fonts.googleapis.com/css?family=Poppins:300,400,500,600,700';
+@import url(../fonts/font-awesome/css/font-awesome.min.css);
```
Comment (romaslmd) [inline app/assets/stylesheets/spacex/style.css]: is this working w/ rails assets pipeline?

File: app/models/device_group.rb [hunk 13]
```diff
@@ -1,5 +1,7 @@
 class DeviceGroup < ApplicationRecord
+	has_many :devices, dependent: :destroy
+  belongs_to :organization
```
Comment (romaslmd) [inline app/models/device_group.rb]: align this, keep your code clean

File: app/models/user.rb [hunk 14]
```diff
@@ -0,0 +1,17 @@
+class User < ApplicationRecord
+  belongs_to :organization
+
+  validates :first_name, :last_name, presence: true, length: { in: 2..50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (romaslmd) [inline app/models/user.rb]: ordering! 
```
class Blah
  constants...

  modules...

  ...
end
```

File: app/views/account/users/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,781 @@
...
+</div><!--/#wrapper -->
+  <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
+  <!--[if lt IE 9]>
+    <script src="assets/script/html5shiv.min.js"></script>
+    <script src="assets/script/respond.min.js"></script>
+  <![endif]-->
+  <!-- 
+  ================================================== -->
+  <!-- Placed at the end of the document so the pages load faster -->
+  <script src="assets/scripts/jquery.min.js"></script>
```
Comment (romaslmd) [inline app/views/account/users/index.html.erb]: use rails helpers to include JS, CSS etc.

File: db/migrate/20190421163541_create_devices.rb [hunk 16]
```diff
@@ -1,9 +1,9 @@
 class CreateDevices < ActiveRecord::Migration[5.2]
   def change
     create_table :devices do |t|
-      t.string :name
-      t.string :device_type
-      t.integer :device_group_id
+      t.string   :name
```
Comment (romaslmd) [inline db/migrate/20190421163541_create_devices.rb]: unnecessary change

File: db/migrate/20190421163638_create_device_groups.rb [hunk 17]
```diff
@@ -1,8 +1,8 @@
 class CreateDeviceGroups < ActiveRecord::Migration[5.2]
   def change
     create_table :device_groups do |t|
-      t.string :name
-      t.integer :organisation_id
+      t.string   :name
```
Comment (romaslmd) [inline db/migrate/20190421163638_create_device_groups.rb]: avoid commits on changes that does not have a real reason

File: db/migrate/20190422154904_create_users.rb (hunk 5 above)
Comment (romaslmd) [reply db/migrate/20190422154904_create_users.rb]: use `t.belongs_to :organization` to create association and all necessary indexes and columns

File: db/migrate/20190422184900_rename_organisation_id_to_organization_id.rb [hunk 18]
```diff
@@ -1,5 +0,0 @@
-class RenameOrganisationIdToOrganizationId < ActiveRecord::Migration[5.2]
-  def change
```
Comment (romaslmd) [inline db/migrate/20190422184900_rename_organisation_id_to_organization_id.rb]: never remove the migration
if it has no use or is invalid - either refactor it or comment the code so it'd be empty

File: spec/factories/device_groups.rb [hunk 19]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :device_group, class: DeviceGroup do
+    name {Faker::Name.name}
+    association :organization
```
Comment (romaslmd) [inline spec/factories/device_groups.rb]: make a space between attributes and association for visual separation

File: spec/models/account/user_spec.rb [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+    it 'is valid with valid attributes' do
+      expect(user.save).to eq(true)
```
Comment (romaslmd) [inline spec/models/account/user_spec.rb]: incorrect test scenario...
prefer
```
before(:each) do
  user.save
end

subject { user }

it { is_expected.to be_persisted } 
```

File: spec/models/account/user_spec.rb [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    end
+    
+    it "is not valid without a first name" do
+      user.first_name = nil
+      expect(user.save).to_not eq(true)
```
Comment (romaslmd) [inline spec/models/account/user_spec.rb]: this example is not quite accurate.. it's not showing the reason why it's not saved
may be different reasons.. try to have each test example to bring some explicit logic about the test subject

File: spec/models/device_group_spec.rb [hunk 22]
```diff
@@ -1,26 +1,25 @@
 require 'rails_helper'
 
 RSpec.describe DeviceGroup, type: :model do
-  subject {
-    described_class.new(name: "Anything", organization_id: "1")
-  }
+  context 'validation tests' do
+  let(:device_group){create (:device_group)}
+  
+    it "is valid with valid params" do
```
Comment (romaslmd) [inline spec/models/device_group_spec.rb]: what params?
try to make an example description more specific

File: spec/rails_helper.rb [hunk 23]
```diff
@@ -1,12 +1,14 @@
...
 require 'spec_helper'
 ENV['RAILS_ENV'] ||= 'test'
 require File.expand_path('../../config/environment', __FILE__)
 # Prevent database truncation if the environment is production
 abort("The Rails environment is running in production mode!") if Rails.env.production?
 require 'rspec/rails'
+require 'support/factory_bot'
 # Add additional requires below this line. Rails is not loaded until this point!
-
+require 'support/factory_bot'
```
Comment (romaslmd) [inline spec/rails_helper.rb]: no need to require multiple times the same file.. merge conflict?

File: spec/spec_helper.rb [hunk 24]
```diff
@@ -93,4 +93,7 @@
   # as the one that triggered the failure.
   Kernel.srand config.seed
 =end
+
+
+
```
Comment (romaslmd) [inline spec/spec_helper.rb]: ?

File: spec/support/factory_bot.rb [hunk 25]
```diff
@@ -0,0 +1,6 @@
+RSpec.configure do |config|
+  config.include FactoryBot::Syntax::Methods
+end
+
+
```
Comment (romaslmd) [inline spec/support/factory_bot.rb]: keep your code clean

File: app/assets/stylesheets/spacex/style.css (hunk 12 above)
Comment (igorkasyanchuk) [reply app/assets/stylesheets/spacex/style.css]: connect as gem (and check what version it's using and find appropriate gem

File: app/controllers/account/users_controller.rb [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+  end
+
+  def edit
+    @user = resource
+  end
+
+  def update
+    @user = resource
+    if @user.update_attributes(user_params)
+      redirect_to account_user_path(id: @user.id)
```
Comment (pereslop) [inline app/controllers/account/users_controller.rb]: you don't need to directly set `id: @user.id`, try something `user: @user`,

File: app/controllers/account/users_controller.rb [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+    else
+      render 'edit'
```
Comment (pereslop) [inline app/controllers/account/users_controller.rb]: try avoid to user strings, when it is not nessasery, use symbols, `render :edit`

File: spec/models/user_spec.rb [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+
+    describe 'factory' do
+      it { is_expected.to be_truthy }
+    end
+
+    describe '#first_name' do
+      it { is_expected.to validate_presence_of(:first_name) }
+
+      it 'does not save the object with 1 letter' do
+        user.update(first_name: "A")
```
Comment (pereslop) [inline spec/models/user_spec.rb]: you don't need to update user in the database, to test if it valid. Use just  `user.first_name = 'A'` and `user.validate` check errors messages

File: spec/models/user_spec.rb [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+        expect(user).not_to be_valid
+        expect(user.errors[:first_name]).to include("is too short (minimum is 2 characters)")
+      end
+    end
+
+    describe '#last_name' do
+      it { is_expected.to validate_presence_of(:last_name) }
+
+      it 'does not save the object with 50+ letters' do
```
Comment (pereslop) [inline spec/models/user_spec.rb]: and you should rename the spec,  because it is not checking, if user is saved in database.

File: spec/models/device_spec.rb (hunk 10 above)
Comment (pereslop) [reply spec/models/device_spec.rb]: add space before curly brackets

File: app/assets/javascripts/spacex/fullcalendar.init.js [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+ * Module/App: Full Calendar
+ */
+
+$(document).ready(function() {
+
+
+	/* initialize the calendar
+	-----------------------------------------------------------------*/
+
+	if ($('#calendar').length){
```
Comment (igorkasyanchuk) [inline app/assets/javascripts/spacex/fullcalendar.init.js]: do we have calendar?
I think no, so this JS could be deleted

File: app/models/user.rb (hunk 14 above)
Comment (igorkasyanchuk) [reply app/models/user.rb]: what about comment above? FIX it

read "rails code style" what should be first, second, ...

File: app/models/user.rb [hunk 31]
```diff
@@ -0,0 +1,21 @@
...
+    presence: true,
+    format: { with: VALID_PASSWORD_REGEX },
+    confirmation: true,
+    on: :create
+  validates :password,
+    allow_nil: true,
+    format: { with: VALID_PASSWORD_REGEX },
+    confirmation: true,
+    on: :update
+  enum role: [:member, :admin, :super_admin]
```
Comment (igorkasyanchuk) [inline app/models/user.rb]: read what is first, second, etc....

association, validations, etc. PUT according to rails style guide

File: app/views/account/users/show.html.erb [hunk 32]
```diff
@@ -0,0 +1 @@
+<p>Show user</p>
```
Comment (igorkasyanchuk) [inline app/views/account/users/show.html.erb]: WHY you have account/users here?
is it was a part of your task? REMOVE IT

File: config/routes.rb [hunk 33]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  namespace :account do
+    resources :users
```
Comment (igorkasyanchuk) [inline config/routes.rb]: was it a part of your task?

CREATE NEW dashboard#index controller in namespace account and output just "Hello <first_name" of current user

File: db/migrate/20190421163541_create_devices.rb [hunk 34]
```diff
@@ -1,9 +1,9 @@
...
+      t.string   :device_type
+      t.integer  :device_group_id
```
Comment (igorkasyanchuk) [inline db/migrate/20190421163541_create_devices.rb]: WHY you have in your PR devices? REMOVE

File: db/migrate/20190426093214_add_index_organization_id_to_device_groups.rb [hunk 35]
```diff
@@ -0,0 +1,5 @@
+class AddIndexOrganizationIdToDeviceGroups < ActiveRecord::Migration[5.2]
+  def change
+    add_index :device_groups, :organization_id
```
Comment (igorkasyanchuk) [inline db/migrate/20190426093214_add_index_organization_id_to_device_groups.rb]: remove

File: db/migrate/20190426093344_add_index_device_group_id_to_devices.rb [hunk 36]
```diff
@@ -0,0 +1,5 @@
+class AddIndexDeviceGroupIdToDevices < ActiveRecord::Migration[5.2]
+  def change
+    add_index :devices, :device_group_id
```
Comment (igorkasyanchuk) [inline db/migrate/20190426093344_add_index_device_group_id_to_devices.rb]: remove

File: spec/helpers/users_helper_spec.rb [hunk 37]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+# Specs in this file have access to a helper object that includes
+# the UsersHelper. For example:
+#
+# describe UsersHelper do
```
Comment (igorkasyanchuk) [inline spec/helpers/users_helper_spec.rb]: do you need this file? I don't. REMOVE

File: spec/controllers/dashboard_controller_spec.rb [hunk 38]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe Account::DashboardController, type: :controller do
+  let!(:user) { build(:user) }
```
Comment (igorkasyanchuk) [inline spec/controllers/dashboard_controller_spec.rb]: must be create(:user)

read about different. why "build"?

File: spec/controllers/users_controller_spec.rb [hunk 39]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe Account::UsersController, type: :controller do
+  let!(:user) { build(:user) }
+
+  describe "GET index" do
+    it "should get index" do
```
Comment (igorkasyanchuk) [inline spec/controllers/users_controller_spec.rb]: same here

File: app/assets/stylesheets/users.scss [hunk 40]
```diff
@@ -0,0 +1,9 @@
+// Place all the styles related to the users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
+
+/*
+*= require font-awesome
```
Comment (igorkasyanchuk) [inline app/assets/stylesheets/users.scss]: чому це файл юзерс?

чому це не в головному файлі

взагалі видали цей файл

Comment (igorkasyanchuk) [reply app/assets/stylesheets/users.scss]: тут нема жодного відношення до юзерів

File: app/views/shared/dashboard/_footer.html.haml [hunk 41]
```diff
@@ -0,0 +1,10 @@
+.main-content
+  %footer.footer
+    %ul.list-inline
+      %li 2016 © NinjaAdmin.
```
Comment (igorkasyanchuk) [inline app/views/shared/dashboard/_footer.html.haml]: чому рік захардкоджений

File: config/routes.rb [hunk 42]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  root 'home#index'
+
+  namespace :account do
+    resources :users, :dashboard
```
Comment (igorkasyanchuk) [inline config/routes.rb]: ЩО ЦЕ ЗА РОУТІНГ ТАКИЙ з дашбордом?
